Test/Integration/Category/CategoryCacheTest.php

```diff
@@ -0,0 +1,341 @@
+            "expected FPC tag on category {$name} id {$categoryId}"
+        );
+
+        if ($hasProducts) {
+            $this->assertMatchesRegularExpression('/<div.*class=.*products-grid.*>/', $response->getContent(), 'Backend content was not rendered.');
+        }
+    }
+
+    /**
+     * @dataProvider getCategoryProvider
+     * @depends      testCategoryPlpMissBackendRenderOn
+     * @magentoConfigFixture current_store system/full_page_cache/caching_application 1
+     * @magentoConfigFixture current_store algoliasearch_advanced/advanced/prevent_backend_rendering 0
+     * @magentoConfigFixture current_store algoliasearch_instant/instant/replace_categories 1
+     * @magentoCache full_page enabled
+     */
+    public function testCategoryPlpHitBackendRenderOn(int $categoryId, string $name): void
+    {
+        $replace = $this->config->getValue('algoliasearch_instant/instant/replace_categories', ScopeInterface::SCOPE_STORE);
+        $this->assertEquals(1, $replace,"Replace categories must be enabled for this test.");
+
+        $this->registerPageHitSpy();
+
+        $this->dispatchCategoryPlpRequest($categoryId);
+        $response = $this->getResponse();
+        $this->assertEquals(200, $response->getHttpResponseCode(), 'Request failed');
+    }
+
+    /**
+     * @dataProvider getCategoryProvider
+     * @depends      testFullPageCacheAvailable
+     * @magentoConfigFixture current_store system/full_page_cache/caching_application 1
+     * @magentoConfigFixture current_store algoliasearch_advanced/advanced/prevent_backend_rendering 1
+     * @magentoConfigFixture current_store algoliasearch_instant/instant/replace_categories 1
+     * @magentoCache full_page enabled
+     */
+    public function testCategoryPlpMissBackendRenderOff(int $categoryId, string $name, bool $hasProducts): void
+    {
+        $preventBackend = $this->config->getValue('algoliasearch_advanced/advanced/prevent_backend_rendering', ScopeInterface::SCOPE_STORE);
+        $this->assertEquals(1, $preventBackend,"Prevent backend rendering must be enabled for this test.");
+
+        $this->resetCache(__METHOD__);
+        $this->dispatchCategoryPlpRequest($categoryId);
+        $response = $this->getResponse();
+        $this->assertEquals(200, $response->getHttpResponseCode(), 'Request failed');
+        $this->assertEquals(
+            'MISS',
+            $response->getHeader('X-Magento-Cache-Debug')->getFieldValue(),
+            "expected MISS on category {$name} id {$categoryId}"
+        );
+        $this->assertContains(
+            'FPC',
+            explode(',', $response->getHeader('X-Magento-Tags')->getFieldValue()),
+            "expected FPC tag on category {$name} id {$categoryId}"
+        );
+
+        if ($hasProducts) {
+            $this->assertDoesNotMatchRegularExpression('/<div.*class=.*products-grid.*>/', $response->getContent(), 'Backend content was rendered.');
```

Comment:
As discussed on slack, this test fails for me for some reason, maybe because I only cache one version on the page ? This may be related to integration test config though

Reply:
It's not failing for me so I suspect it's the test throwing off a false positive on the RegEx. To confirm we'd need to inspect the content you're seeing on your local. I'm suspecting it might be injecting a different type of `products-grid` that is not related to the "real" `products-grid` I'm looking for. Once we determine that we can fix the test but since I can't reproduce I'll leave it for now.